File: 2-add-two-numbers/add-two-numbers.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(0);
        ListNode temp = dummy;
        int carry = 0;
        while(l1 != null || l2 != null)
        {
            int rem = 0;
            if(l1 == null)
            {
                rem = (l2.val + carry) % 10;
                carry = (l2.val + carry) / 10;
                temp.next = new ListNode(rem);
                l2 = l2.next;
            }
            else if(l2 == null)
            {
                rem = (l1.val + carry) % 10;
                carry = (l1.val + carry) / 10;
                temp.next = new ListNode(rem);
                l1 = l1.next;
            }
            else
            {
                rem = (l1.val + l2.val + carry) % 10;
                carry = (l1.val + l2.val + carry) / 10;
                temp.next = new ListNode(rem);
                l1 = l1.next;
                l2 = l2.next;
            }
            temp = temp.next;
        }

        if(carry > 0)
        {
            temp.next = new ListNode(carry);
        }

        return dummy.next;
    }
}

File: 2-add-two-numbers/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        return Add(l1, l2, 0);
    }

    ListNode* Add(ListNode* l1, ListNode* l2, int carry) {
        if (l1 == nullptr && l2 == nullptr && carry == 0)
            return nullptr;

        int val1 = (l1 != nullptr) ? l1->val : 0;
        int val2 = (l2 != nullptr) ? l2->val : 0;

        int sum = val1 + val2 + carry;
        int newDigit = sum % 10;
        int newCarry = sum / 10;

        ListNode* node = new ListNode(newDigit);
        node->next = Add((l1 != nullptr) ? l1->next : nullptr,
                         (l2 != nullptr) ? l2->next : nullptr, newCarry);

        return node;
    }
};